tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+    auto tensorCur_type = tensor_val.getType().dyn_cast<RankedTensorType>();
+    if (!tensorCur_type) return {};
+    tensor_shape_vec.push_back(tensorCur_type.getShape());
+  }
+  SmallVector<Value, 2> dyn_sizes;
+  for (std::string const& dim_ind : output_loop_vec) {
+    int arg_number = 0;
+    auto dim_ind_it =
+        std::find(lhs_loop_vec.begin(), lhs_loop_vec.end(), dim_ind);
+    auto dim_ind_pos = dim_ind_it - lhs_loop_vec.begin();
+    if (dim_ind_it == lhs_loop_vec.end()) {
+      arg_number = 1;
+      dim_ind_it = std::find(rhs_loop_vec.begin(), rhs_loop_vec.end(), dim_ind);
+      dim_ind_pos = dim_ind_it - rhs_loop_vec.begin();
+    }
+    if (tensor_shape_vec[arg_number][dim_ind_pos] != ShapedType::kDynamicSize)
+      continue;
+    dyn_sizes.push_back(
+        b.create<memref::DimOp>(loc, tensor_vec[arg_number], dim_ind_pos));
+  }
+  return dyn_sizes;
+}
+
+// Adds indices/axes that are missing from output set
+std::set<std::string> FindSummationAxes(std::set<std::string> input_set,
+                                        std::set<std::string> output_set) {
+  std::set<std::string> summation_axes;
+  for (auto ind : input_set) {
+    if (output_set.find(ind) == output_set.end()) summation_axes.insert(ind);
+  }
+  return summation_axes;
+}
+
+// Given a 1:1 map from std::string -> affine dimension expression
+// we can get the affine expression of dimensions that an
+// operand will access based on the input_str of einsum_config
+// For example:
+// let string_dim_umap = {'a' : d0, 'b' : d1, 'c' : d2}
+// for einsum_config "abc,cb->acb"
+// the first_input_operand will get umap[{"a","b","c"}] -> (d0, d1, d2)
+// the second_input_operand will get umap[{"c","b"}] -> (d2, d1)
+// the ouput_operand will get umap[{"a","c","b"}] -> (d0, d2, d1)
+SmallVector<AffineExpr, 4> getExprFromConfig(
+    std::set<std::string> input_set, std::vector<std::string> loop_dims,
+    std::unordered_map<std::string, AffineExpr> str_affine_dim_umap) {
+  SmallVector<AffineExpr, 4> OutputExpr;
+  for (std::string const& dim : loop_dims) {
+    OutputExpr.push_back(str_affine_dim_umap[dim]);
+  }
+  return OutputExpr;
+}
+
+// Given a real number n, will generate string of: "012...(n-1)"
+std::string rangeStr(size_t number) {
+  std::string range_str;
+  for (size_t i = 0; i < number; i++) range_str.append(std::to_string(i));
+  return range_str;
+}
+
+// Convert mhlo.einsum op into linalg.generic
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    // Find maximum rank / number of loops.
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config().str();
```

Comment:
Why the `.str()`? Please avoid copying string around unnecessarily: try to use as much StringRef as possible everywhere.

Reply:
done